Monkey patching in ruby (#178)
incubyte/incubyte.github.io: content/blog/monkey-patch.md

+++
title = "Monkey Patching"
slug = "monkey-patch"
date = 2023-10-17T18:29:54+05:30
image = "/images/2023/10/monkeywithsword.jpeg"
draft = false
authors = ["Nitin Rajkumar"]
description = "Monkey patching and it's sideeffects in Ruby"
tags = ["Software Craftsmanship", "GraphQL"]
categories = ["Software Craftsmanship", "Ruby", "Monkey Patching"]
type = ""
+++

Ruby, being a dynamic programming language provides the ability and hence freedom to reopen the existing classes and change their behavior during runtime by adding new methods or modifying existing methods. 

#### How to monkey patch

> ```ruby

Reviewer: The indentation in this codeblock can be fixed

> class Sound
> 	def car
> 		"honk"
> 	end
> end
> class Sound
> 	def bike
> 		"peep-peep"
> 	end
> end

Reviewer 2: use space instead of tab for indentation and use 2 space indentation. it will look better and will also align with indentation used at line 43 to 47
```suggestion
class Sound
  def car
    "honk"
  end
end
class Sound
  def bike
    "peep-peep"
  end
end
```

> class Sound
>   def car
> 		"honk"
> 	end
> end
> class Sound
> 	def bike
> 		"peep-peep"
> 	end
> end

Reviewer: ```suggestion
class Sound
    def car
		"honk"
	end
end
class Sound
	def bike
		"peep-peep"
	end
end
```

> noise = Sound.new
> noise.car #=> honk
> noise.bike #=> peep-peep

Reviewer 2: Either rename the class to Noise or rename the variable to sound

Noise is a special category of Sound
Not all sounds are noise

so Sound.new cannot be called a noise
```suggestion
sound = Sound.new
sound.car #=> honk
sound.bike #=> peep-peep
```

```

> The above is a simple example where we reopened Sound class and added a new method. It provides many other opportunities including overriding the method of the original class.

Reviewer: Would it add value to mention that the Sound class lre existed in a library or gem and then present these methods as patches?

Author: We can also reopen the classes from our own code base. The other example in this blog mentions overriding String class.

Author: It might make it convoluted.

Reviewer: We should mention the Sound class is a pre existing one, right now it will seem to be an entirely new class and will not convey monkey patching, the String.upcase example makes sense immediately

#### When does one think of using it
Usually when the programmer wants to fix the bug quickly.
1. When the gem or library that is being used has a bug and before a patch-fix from the library if one wants to fix it in their code base.
2. When certain method you expect the library to contain is missing one can monkey patch the method.

> #### Side Effects

Reviewer 2: Point number 1,2,3,4 on line 50 to 53 all talk about negative effects of using Monkey Patching

The word Side effects does not mean negative Side effects by default

May be the word "Cons" suits better here

```ruby
class String
  def upcase
    # do nothing 
  end
end
```
The above monkey patch would return nothing causing the `upcase` to fail everywhere the patch is being used.
1. In a large code base multiple developers may be monkey patching the same method in different ways.
2. When a 3rd-party gem is monkey patched, upgrading the gem may result in conflict.
3. Can be difficult to track which patch is giving the result.
4. Adds to tech debt as monkey patching needs to be cleaned out once the fix is in place.
#### Measures to take while monkey-patching
 
1. Instead of reopening the classes where needed, it is considered a good idea to have the patch in module and include it in the class.
2. Document the applied monkey patches and clean them up when the fix is in place.

#### Conclusion

> Be a good Samaritan and contribute to the open source. When a bug is discovered in 3rd party library raise a issue, fix it, and get the pull request merged meanwhile add a monkey patch. Monkey patching has gained the popularity as the power tool in ruby because of the havoc it can cause if it is misused. Refrain from using it and use it only when it is the last resolution.

Reviewer 2: remove extra commas and add more commas

if possible , run grammarly on this paragraph and autofix this

```suggestion
Be a good Samaritan and contribute to the open source.
When a bug is discovered in 3rd party library: raise a issue, fix it and get the pull request merged, meanwhile add a monkey patch.
Monkey patching has gained popularity as the power tool in ruby because of the havoc it can cause if it is misused.
Refrain from using it and use it only when it is the last resolution.
```

+++
title = "Monkey Patching"
slug = "monkey-patch"
date = 2023-10-17T18:29:54+05:30
image = "/images/2023/10/monkeypatch.jpg"
draft = false
authors = ["Nitin Rajkumar"]
description = "Monkey patching and it's sideeffects in Ruby"
tags = ["Software Craftsmanship", "GraphQL"]
categories = ["Software Craftsmanship", "Ruby", "Monkey Patching"]
type = ""
+++

Ruby, being a dynamic programming language provides the ability and hence freedom to reopen the existing classes and change their behavior during runtime by adding new methods or modifying existing methods. 

#### How to monkey patch

```ruby
class Sound
  def car
    "honnnnnk..."
  end
end

class Sound
  def bike
    "peep-peep"
  end
end
sound = Sound.new
sound.car #=> honnnnnk...
sound.bike #=> peep-peep
```
The above is a simple example where we reopened Sound class and added a new method. It provides many other opportunities including overriding the method of the original class.

> 

Reviewer 2: #### Another example

```ruby

# Original definition of Vehicle class which only handles cars
class Vehicle
  def initialize(kind)
    @kind = kind
  end

  def horn
    if @kind == 'car'
      "honnnnnk..."
    else
      "tirrr"
    end
  end
end

vehicle = Vehicle.new("car")
vehicle.horn #=> honnnnnk...

vehicle = Vehicle.new("bike")
vehicle.horn #=> peep-peep

vehicle = Vehicle.new("truck")
vehicle.horn #=> peep-peep

# Overriding horn method of Vehicle class to support Trucks also
class Vehicle
  def horn
    if @kind == 'car'
      "honnnnnk..."
    elsif @kind == 'truck'
      "loud-honk"
    else
      "tirrr"
    end
  end
end

vehicle = Vehicle.new("car")
vehicle.horn #=> honnnnnk...

vehicle = Vehicle.new("bike")
vehicle.horn #=> peep-peep

vehicle = Vehicle.new("truck")
vehicle.horn #=> loud-honk
```

Author: The example in Cons covers this.

#### When does one think of using it

> Usually when the programmer wants to fix the bug quickly.
> 1. When the gem or library that is being used has a bug and before a patch-fix from the library if one wants to fix it in their code base.
> 2. When certain method you expect the library to contain is missing, one can monkey patch the method.

Reviewer 2: Improved grammer: Credits ChatGpt

```suggestion
Usually, when a programmer wants to quickly fix a bug, it occurs under the following circumstances:

1. When the gem or library being used has a bug and there is no patch or fix available from the library yet, one may wish to fix it in their own codebase.

2. When a certain method, which you expect the library to have, is missing, one can apply a monkey patch to that method.
```

#### Cons

```ruby
class String
  def upcase
    # do nothing 
  end
end
```

> The above monkey patch would return nothing causing the `upcase` to fail everywhere the patch is being used.
> 1. In a large code base multiple developers may be monkey patching the same method in different ways.
> 2. When a 3rd-party gem is monkey patched, upgrading the gem may result in conflict.
> 3. Can be difficult to track which patch is giving the result.
> 4. Adds to tech debt as monkey patching needs to be cleaned out once the fix is in place.

Reviewer 2: Grammar improvement by Chat GPT

```suggestion
The above monkey patch would return nothing, causing the upcase method to fail wherever the patch is applied.

1. In a large codebase, multiple developers may be monkey patching the same method in different ways.

2. When a third-party gem is monkey patched, upgrading the gem may result in conflicts.

3. It can be difficult to determine which patch is producing the desired result.

4. This adds to technical debt, as monkey patches need to be removed once the fix is in place.
```

#### Measures to take while monkey-patching
 
1. Instead of reopening the classes where needed, it is considered a good idea to have the patch in module and include it in the class.
2. Document the applied monkey patches and clean them up when the fix is in place.

> #### Conclusion
> 
> Be a good Samaritan and contribute to open source.
> When a bug is discovered in a third-party library, raise an issue, fix it, and get the pull request merged. Meanwhile, add a monkey patch.
> Monkey patching has gained the popularity it has as a power tool in Ruby because of the havoc it can cause if misused. Refrain from using it and use it only when it is the last resort.

Reviewer 2: Improved wordings

```suggestion
#### Conclusion

Be a good Samaritan and contribute to open source. When you discover a bug in a third-party library, raise an issue, fix it, and get the pull request merged. In the meantime, consider adding a monkey patch.

Monkey patching has gained popularity as a powerful tool in Ruby, but it should be used with caution due to the potential havoc it can cause if misused. Refrain from using it and resort to it only as a last option.
```
